### ISSUE TITLE
Dealing with namespaces and cache files is very confusing (as a consequence of lazy evaluation)

### ISSUE THREAD

**Issue by slowglow** [acoular/acoular#67]
As someone coming from procedural programming, I find it very confusing to create a function (or a set of functions), which create a beamformer and then loop through a list of filenames to apply it to.

Conceptually, I would like to achieve:
```
my_beamformer = create_my_beamformer()
for input_file in file_list:
     output_file = my_beamformer(input_file)
```

The catch is that instead of converting all my measurement files to *.h5, I want to convert them one by one to a temporary *.h5 file, which then gets overwritten every time I call the beamformer function with a newly converted measurement file.

I tried two approaches.

*Approach 1*

```
def create_temp_h5_file(h5fname_out, data, fs, series_title):
    import tables

    acoularh5 = tables.open_file(fh5fname_out, mode="w", title=series_title)
    # Note that we create 'EArray'
    acoularh5.create_earray('/', 'time_data', atom=None, title='',
                            filters=None, expectedrows=100000,
                            chunkshape=[256, 64], byteorder=None,
                            createparents=False, obj=data)
    acoularh5.set_node_attr('/time_data', 'sample_freq', fs)
    acoularh5.close()  # <- here the *.h5 file is closed after creation
    return

def create_my_beamformer(layout_file, datafile)
    mg = acoular.MicGeom(from_file=layout_fullfile)
    data, fs, series_title = read_from_datafile(datafile)  # <- pseudocode to return the data
    create_temp_h5_file(temp_h5_datafile, data)
    ts = acoular.TimeSamples(name=temp_h5_datafile)  # This leaves the .h5 file open, despite closing it at creation
    ts.calib = acoular.Calib(from_file=calib_fullfile)

    ps = acoular.PowerSpectra(time_data=ts, block_size=128, window='Hamming')
    rg = acoular.RectGrid(x_min=-0.2, x_max=0.2, y_min=-0.2, y_max=0.2, z=0.235,
                          increment=0.01)
    st = acoular.SteeringVector(grid=rg, mics=mg)
    bb = acoular.BeamformerBase(freq_data=ps, steer=st)
    return bb
```
With this approach, the returned beamformer knows nothing of the TimeSamples object ts on which it operates by lazy evaluation. 
In order for it to work, do I need to return everything to the caller: ts, ps, st, rg, bb?
If that is the case putting it in a separate function is not very meaningful.  So, I tried

*Approach 2*

In a second approach, I tried putting everything in one big function, which:
1. takes as input argument only the filename (of a file, which is not *.h5)
2. converts to temporary *.h5 file
3. creates ts, ps, st, rg, bb
4. evaluates the result of bb

On the first filename the function works OK, on the second run I get 
```
'H5CacheFileTables' object has no attribute 'root'
```
In this approach cleaning-up the cache files is impossible, unless I restart the kernel.
I tried to find the blocking process using:

```
proc = psutil.Process()
open_fname_list = proc.open_files()
for opfile in open_fname_list:
    if Path(opfile.path).suffix == '.h5':
        print(opfile.path)
        opfile.kill()
        os.close(opfile.fd)
```
however, although I can find the open files, I cannot close them, because they are blocked by some process.
I also can't find the blocking process to kill it. 
Finally, I am not sure killing the process is the best course of action.

Any advice?

**Comment by slowglow** [comment]
OK, with some tedious debugging, it seems that with Approach 2, the second time the function is called 
```
ts = acoular.TimeSamples(name="temp.h5")
```
acoular does not create a new `temp_cache.h5` file, presumably, because the `temp.h5` filename is the same. 
I expect the `ts` object to live inside the function namespace, so the second time I call the function it has forgotten about the filename, which I passed the first time. Moreover, adding `del ts` and deleting the 'temp.h5' and 'temp_cache.h5' files with `os.remove` does not seem to solve the problem. So, where does the `ts` object live and more generally, how do I use acoular inside my own functions?

Am I missing something in regard to the use of namespaces?

**Comment by esarradj** [comment]
The problem with both your approaches is that you change the content of the temporary .h5 file while you do not change the name of it (that's what I guess). For the naming of cache files, Acoular assumes that an .h5 input file that has a certain name has a certain (immutable) content.
I see three possible solutions: 
- give your temporary input a new name each time you write different data into it.
- write a replacement class for `TimeSamples` that can handle your kind of input data (refer to the code, you will need to implement the `SamplesGenerator` class)
- switch of all caching via `acoular.config.global_caching = 'none'` (this means that you will have to repeat the computation in case you have the same data)

**Comment by esarradj** [comment]
If you return a `BeamformerBase` as in your first approach, it will know everything about its data source, Mic geometry and so on, and you can ask about it easily, e.g. `bb.freq_data.time_data.sample_freq`. Therefore it is sufficient to return just `bb` and not all the other stuff. However, if you call the function again, you will get a different `BeamformerBase` object.

**Comment by slowglow** [comment]
Hello and thank you for your response.

As a temporary workaround, I implemented your suggestion to give a new name to the temporary input. 
It works flawlessly. 

Replacement class is probably technically superior. For the time being switching the caching off is also sufficient for my purpose.

Regarding the return of `BeamformerBase` from the function, I struggled last week, possibly, because I did not know which property to tweak, or maybe, because the temp filename was the same it did not have effect on the beamformer output. 

Would it be possible to document all of the tips you gave into a section in the documentation on caching and lazy evaluation?
